### ISSUE TITLE
Ticket T12：BRATS 目标几何（1mm 各向同性 256×256×128，crop+pad 替代 resize）

### ISSUE THREAD

**Issue by ACautomata** [ACautomata/NV-Generate-CTMR#39]
Part of #13（BRATS 微调 rflow-mr-brain 训练 spec）。本 spec **修订 #13 的 §②.5 / §⑤.1 / §⑤.3**：BRATS 全链路改用**目标几何**（1mm 各向同性、`dim [256, 256, 128]`、中心对齐），MR-RATE replay、上游 configs 与上游 scripts 全部不动。词汇遵循 `CONTEXT.md`（目标几何 / spacing / dim / FOV / 条件 / latent / replay / 遗忘参照集 / 冻结基线），决定记录见 `docs/adr/0001-brats-finetune-target-geometry.md`。

## Problem Statement

现在微调出来的 BraTS 模型，生成的图像达不到「可以拿给医生看」的标准：同一份权重在 1×1×1 条件下采样明显更清楚，而在现行管线的 `[0.9375, 0.9375, 1.2109]` 条件下采样偏糊。

根因是几何而不是别的：现行管线把 240×240×155 @1mm 的原始 volume **直接三线性 resize** 到 `dim [256, 256, 128]`，spacing 随之被改写成 `[0.9375, 0.9375, 1.2109]`。这里同时坏了两件事：

1. resize 对原始体素做了插值重采样，原始 1mm 的采样网格被破坏；
2. 训练/生成用的几何与 FID 评测协议用的几何**不是同一个域**——评测协议早就把真实图像重采样到 1mm、中心裁剪到 256×256×128、再对称 padding（`Spacingd(1mm)` → `CenterSpatialCropd(256×256×128)` → `SpatialPadd`），也就是说**真实侧本来就落在 1mm 的 256×256×128 上**，只有模型这一侧不在。

结果是训练域、生成域、评测域三处各说各话；而 FOV 是这类模型最敏感的质量旋钮，三处不一致时，FID 数字与目视质量都失去参照。

## Solution

在 latent 编码**之前**插入一个 BraTS 专属的独立前置 stage，把每个 case 从原生 240×240×155 @1mm **中心地**转成目标几何：

- 平面两个轴：pad 到 256（每边 8 层）——实测这部分落在脑组织之外的空白里，无损；
- 头脚轴：中心裁掉底部 13 层、顶部 14 层（155 → 128）；
- spacing 全程保持 1×1×1，**不做任何插值重采样**。

结果落盘到镜像原相对路径的目录树，再交给**零修改**的上游编码管线；上游的 `Resized` 到 `[256, 256, 128]` 在这时是数学恒等变换。此后训练、生成、评测三处共用同一个几何，且该几何**逐体素等于 FID 协议落到真实图像上的那块区域**——「训练域 = 评测域」不再是声明，而是一条可执行的断言。

MR-RATE replay、旧 label 的遗忘检查、上游 configs 与上游 scripts 保持 v1 原生条件不变：本 spec 只对 BraTS 有特殊要求。

## User Stories

1. 作为一个要给医生看 BraTS 生成结果的人，我想要生成图是 1×1×1 各向同性，这样它在影像上是可判读的。
2. 作为一个跑微调的人，我想要 BraTS 的训练、生成、评测三处共用同一个目标几何，这样 FID 数字与目视质量指向同一件事。
3. 作为一个实现者，我想要几何由**中心裁剪 + 对称 padding** 定义而不是 resize，这样原始 1mm 采样网格不被插值破坏。
4. 作为一个实现者，我想要前置 stage 是一个独立脚本 + 独立输出目录，这样上游的编码与训练代码一个字都不用改。
5. 作为一个实现者，我想要几何常量（`dim`、`spacing`、裁剪层数、padding 值）只在一处定义、其余地方引用它，这样各处不会漂移。
6. 作为一个审查者，我想要一条测试断言新几何与 FID 协议的重采样链**逐体素一致**，这样「训练域 = 评测域」是被执行的事实而不是文档里的声明。
7. 作为一个审查者，我想要目标几何通过 `check_input_mr` 的硬约束，这样生成阶段不会被校验器拒绝（否则验收时会以「输入非法」这种误导性的方式失败）。
8. 作为一个重跑者，我想要已存在的输出文件被跳过，这样中断之后重跑不重算。
9. 作为一个组装混合 dataset.json 的人，我想要输出树镜像输入的相对路径，这样阶段①的 dataset.json 原样可用、不需要第二个 dataset.json。
10. 作为一个跑混合训练的人，我想要 BraTS 条目在合并后的 dataset 里的路径仍然解析得到，这样混合训练不会静默少样本。
11. 作为一个跑遗忘检查的人，我想要 MR-RATE replay 与 10 个旧 label 的生成条件保持 v1 原生 `[0.94, 0.94, 1.36]`，这样与冻结基线仍可比。
12. 作为一个跑 FID 的人，我想要评测工具的真实侧协议完全不动，因为它已经落在目标几何上。
13. 作为一个运维者，我想要**旧几何的 latent 被显式拒绝**而不是静默进入训练，这样几何切换不会无声失败。
14. 作为一个运维者，我想要几何一旦冻结，改几何就必须换输出目录（而不是覆盖旧产物），这样历史 run 可追。
15. 作为一个跑基线的人，我想要 T8 的 pretrained 冻结基线按 `[1, 1, 1]` 重新生成并重新冻结，这样对照臂与微调臂同几何。
16. 作为一个跑验收的人，我想要 BRATS 的 real-real FID 地板不用重算，因为真实侧本来就落在目标几何上（由故事 6 的断言担保）。
17. 作为一个 debug 的人，我想要脚本打印转换前后的 `dim` / `spacing` 与写入/跳过计数，这样一次跑批的结果可自证。
18. 作为一个关心「resize 破坏细节」这件事的人，我想要脚本报告**有多少 case 原生不是 1mm、真的被重采样了**，这样「不插值」这个说法在运行时是可证伪的。
19. 作为一个维护者，我想要上游 `configs/` 与上游 `scripts/` 保持零改动，这样本项目与上游的 diff 仍然只落在 BraTS 相关的新文件与 BRATS 配置上。
20. 作为一个维护者，我想要 BRATS 专用推理配置与 upstream 的差异被逐一断言（只允许几何字段不同），这样日后的手改不会悄悄带走别的字段。
21. 作为一个未来的读者，我想要这套决定的「为什么」留在 ADR 与词汇表里（目标几何已在 `CONTEXT.md` 登记），这样六个月后不会有人把它当 bug 改回去。
22. 作为一个验收者，我想要在 gauss 上用真实数据抽查重编码后的 latent 的 shape 与 spacing，这样落盘产物本身被验证过而不只是代码被验证过。
23. 作为一个给医生出 case 的人，我想要训练完成后把典型样例发到 GitHub Release，这样评审发生在图像域而不是数字域。
24. 作为一个实现者，我想要这次的几何变更**不删除任何既有产物**（旧 latent、旧基线），只在新的 run dir 里产出新的一套，这样回退不需要重新下载或重算任何东西。

## Implementation Decisions

**几何定义（唯一的几何事实来源）**

新增一个几何策略模块，把「目标几何」建模成一个领域概念（`CONTEXT.md` 已登记该词条）。它暴露 `dim`、`spacing`，以及一条把原生 volume 变成目标几何的 MONAI 变换链。变换链的**顺序本身就是决定**，原型产出的组合如下（保留原型片段因为顺序比散文更精确）：

```
LoadImage → EnsureChannelFirst → Orientation(RAS) → float32
  → Spacingd(1mm, bilinear)
  → CenterSpatialCropd([256, 256, 128])
  → SpatialPadd([256, 256, 128], mode="constant", value=0)
```

- **顺序与 FID 协议一致**（裁剪在 padding 之前），这正是故事 6 那条断言能成立的原因。
- 先 `Orientation(RAS)` 再谈「中心」与「z」：否则不同 case 的轴含义不一致。
- 保留 `Spacingd(1mm)` 是让「1mm」成为**不变式**而不是假设；在 BraTS GLI 原生数据上它是恒等变换（抽查 header 为 `[1.0, 1.0, 1.0]`），故事 18 要求把「有多少 case 真的被重采样」报出来。
- `CenterSpatialCropd` 在 roi 大于该轴长度时按 MONAI 语义取满该轴（240 < 256 的平面轴因此原样留下），padding 默认 `method="symmetric"` 对称补边，两者合起来正好给出「每边 8 层」与「底 13 / 顶 14」。
- **强度归一化不在这个 stage 里做**，仍由上游编码管线按 modality 分派。FID 的 compose 有一句告诫：MR 的百分位统计必须在 padding 之前算，否则 padding 体素会稀释分位数。这里不冲突，原因是 BraTS 是 skull-stripped、背景**精确为 0**：pad 只是把更多 voxel 放进分布最底部的那个 bin，(0, 99.5) 两个锚点中下限仍是 0、上限只在非零体素里移动极小的一档。这个理由要写进模块 docstring，否则未来的读者会以为这里违反了上游那条告诫。

**落盘驱动（新的薄驱动脚本）**

- 输入：阶段①的 dataset.json、源 `data_base_dir`、输出根；几何来自上面的策略模块。
- 输出：`输出根 / <dataset.json 里的 image 字符串>`——**镜像相对路径**，于是阶段①的 dataset.json 可以原样复用，不需要第二个 dataset.json（故事 9）。
- 已存在则跳过（上游编码脚本的同一惯例），计数并打印。
- 保留源 NIfTI 的 dtype（原数据是 int16），不引入精度变化。
- 不写 seg：seg 不进 dataset.json、不参与训练。
- 不触碰上游 `scripts.diff_model_create_training_data.py` 与 `scripts/transforms.py`（故事 4、19）。

**latent sidecar 增加几何契约**

- 现有的 sidecar 写入器（BRATS 与 replay 共用）增加一个可选的「要求 spacing 等于」参数。
- BRATS 阶段②传入目标几何的 spacing，于是**任何旧几何的 latent 集合会被拒绝，且拒绝发生在写任何 sidecar 之前**（故事 13）。replay 侧不传，行为不变。
- 这条把 ADR 0001 里「已编码的 BraTS latent 全部作废」从一句提醒变成一道闸门。

**BRATS 专用推理配置**

- 只改 `config_maisi_diff_model_rflow-mr-brain-brats.json` 的 `diffusion_unet_inference.spacing` → `[1, 1, 1]`（`dim` 已经是 `[256, 256, 128]`）。
- 上游 `config_maisi_diff_model_rflow-mr-brain.json` 一个字都不改。
- 这条**修订 #13 §⑤.1 与 §⑤.3**：§⑤.1 要求该推理块与 upstream 逐字节相同，而 §⑤.3 又建议 BRATS 微调模型的推理 spacing 取 `[0.94, 0.94, 1.21]`，两者本就无法同时成立。现在的划分是——**BraTS 生成条件**住在 `-brats` 配置里（目标几何），**遗忘检查的原生 v1 条件**由 upstream 配置提供，两者互不覆盖。

**作废与重算（不删除任何东西）**

- 已编码的 BraTS latent（旧几何）作废，重编码进新的 run dir。
- T8 的 pretrained 冻结基线必须按 `[1, 1, 1]` 重新生成并重新冻结（旧基线在几何变更后不可比）；BRATS 的 real-real FID 地板不用重算。
- 旧产物保留在盘上，不删。

## Testing Decisions

好的测试只断言**外部行为**：给定一个输入 volume，输出 volume 的几何与体素内容是什么；文件落在哪个路径；什么情况下被拒绝。不断言内部实现细节——不调私有方法、不比对 transform 列表的形状、不检查配置对象的字段名。

用户已确认采用 **4 个接缝**：

1. **几何策略模块（新，最高点）。** 用合成的 240×240×155 @1mm volume 驱动：断言输出的 `dim`/`spacing`；断言内容窗口（平面每边 8 层是 padding、头脚轴取的是原 volume 的 `[13, 141)`，用一个可定位的标记值验证而不是只看形状）；断言与 FID 协议那条组合（`ModalityPreprocessing.compose` 以同样的 target_shape / resample_spacing / center_crop / pad 参数构造）**逐体素相等**；断言该几何通过 `check_input_mr`。
2. **落盘驱动（新）。** 用假 dataset.json + 假 volume：断言输出路径恰好是 `输出根 / image 字符串`；断言重跑跳过已存在文件但补齐缺失文件；断言计数。
3. **sidecar 的 1mm 契约（扩展现有 sidecar 写入器）。** 断言 spacing 不符时抛错**且零 sidecar 落盘**；断言 replay 侧不传契约时行为不变。
4. **BRATS 推理配置（扩展现有静态断言）。** 断言 `-brats` 与 upstream 的推理块逐字段相同、**唯 `spacing` 不同**，并且差异恰好是 `[1, 1, 1]`。

Prior art：`tests/test_create_brats_sidecars.py`（假数据 layout + 一个 `subprocess` 端到端用例的写法）、`tests/test_brats_finetune_configs.py`（静态 config 断言）、`tests/test_create_brats_dataset_json.py`（用 `@dataclass(frozen=True)` 建模输入输出）。

## Acceptance Criteria

- [ ] 几何策略模块的 4 组断言全绿（含与 FID 协议逐体素一致、过 `check_input_mr`）
- [ ] 落盘驱动、sidecar 契约、推理配置三个接缝的测试全绿；全仓 `pytest` 无回归
- [ ] 在 gauss 上用真实数据跑完整前置 stage，报告「原生即 1mm / 被重采样」的 case 数
- [ ] 抽查重编码后的 latent：shape `(256, 256, 128)`、spacing `[1.0, 1.0, 1.0]`
- [ ] 用旧几何的 latent 集合跑阶段② sidecar，确认被拒绝且未写出任何 sidecar
- [ ] 混合 dataset.json 的审计通过（BraTS 条目在新输出根下全部解析得到）
- [ ] T8 pretrained 基线按 `[1, 1, 1]` 重新生成并重新冻结，抽样名单与条目数与旧基线一致
- [ ] 生成若干 BraTS 样例并目视确认可用（1×1×1 vs 旧条件的对照）

## Out of Scope

- MR-RATE replay 与遗忘参照集的任何改动（它们是 v1 原生的对照组）。
- 上游 `configs/` 与上游 `scripts/` 的改动——包括「顺手把 resize 也改了」。
- 训练侧 transform、`h5` 数据链、`check_input_mr` 校验器本身。
- seg 的目标几何化（seg 不进 dataset.json、不参与训练）。
- FID 工具的改动（真实侧协议不动，因为无需改）。
- 换一种几何用更少解剖损失（`dim_z=256` 的全 padding 方案、只 pad 平面不动 z 的方案）——已在 ADR 0001 里作为 rejected 备选记录。
- 训练执行本身、ckpt 选点、最终发布（T9/T10/T11）。

## Further Notes

- **已知代价**（ADR 0001 已记录）：头脚轴中心 128 层会切到脑组织；z 向 FOV 降到 128mm，低于 MR-RATE 准入下限 140mm，属条件外推——其表现以生成样例验收，这是有意用 FOV 换清晰度。
- **修订清单**：#13 §②.5 的「零修改 + 实得 spacing `[0.9375, 0.9375, 1.2109]`」不再成立（BRATS 侧改为前置 stage + 目标几何，上游代码仍零修改）；§⑤.1 的推理块逐字节相同要求收窄为「除几何字段外相同」；§⑤.3 的推理 spacing 建议由 `[0.94, 0.94, 1.21]` 改为 `[1, 1, 1]`。
- **运维顺序**：前置 stage → 重编码 latent（新 run dir）→ 写 sidecar（此时几何契约生效）→ 按 `[1, 1, 1]` 重新冻结 T8 基线 → 合并 dataset.json → 训练。
- 验收仍以 FID 为主指标，但本次改动的第一验收面是**图像可判读性**；训练结束后按 `skills/publish_experiment.md` 把典型样例发到 GitHub Release。
